File: yyxarch/src/main/java/com/yyxnb/yyxarch/livedata/ObservableLceeLiveData.java
package com.yyxnb.yyxarch.livedata;

import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;

import com.yyxnb.yyxarch.bean.Lcee;
import com.yyxnb.yyxarch.http.RxHttpUtils;
import com.yyxnb.yyxarch.http.exception.ApiException;

import java.lang.ref.WeakReference;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

class ObservableLceeLiveData<T> extends LiveData<Lcee<T>> {
    private WeakReference<Disposable> mDisposableRef;
    private final Observable<T> mObservable;
    private final Object mLock = new Object();

    ObservableLceeLiveData(@NonNull final Observable<T> observable) {
        mObservable = observable;
    }

    @Override
    protected void onActive() {
        super.onActive();

        mObservable.subscribe(new Observer<T>() {
            @Override
            public void onSubscribe(@io.reactivex.annotations.NonNull Disposable d) {
                // Don't worry about backpressure. If the stream is too noisy then
                // backpressure can be handled upstream.
                synchronized (mLock) {
                    mDisposableRef = new WeakReference<>(d);
                    RxHttpUtils.addDisposable(mDisposableRef.get());
                }
                postValue(Lcee.<T>loading());
            }

            @Override
            public void onNext(@io.reactivex.annotations.NonNull T t) {
                if (null == t) {
                    postValue(Lcee.<T>empty());
                } else {
                    postValue(Lcee.content(t));
                }
            }

            @Override
            public void onError(@io.reactivex.annotations.NonNull Throwable t) {
                synchronized (mLock) {
                    RxHttpUtils.cancelSingleRequest(mDisposableRef.get());
                    mDisposableRef = null;
                }
                // Errors should be handled upstream, so propagate as a crash.
                postValue(Lcee.<T>error(ApiException.handleException(t).getMessage()));
            }

            @Override
            public void onComplete() {
                synchronized (mLock) {
                    RxHttpUtils.cancelSingleRequest(mDisposableRef.get());
                    mDisposableRef = null;
                }
            }
        });

    }

    @Override
    protected void onInactive() {
        super.onInactive();
        synchronized (mLock) {
            WeakReference<Disposable> subscriptionRef = mDisposableRef;
            if (subscriptionRef != null) {
                Disposable subscription = subscriptionRef.get();
                if (subscription != null) {
                    subscription.dispose();
                }
                RxHttpUtils.cancelSingleRequest(mDisposableRef.get());
                mDisposableRef = null;
            }
        }
    }
}

File: yyxarch/src/main/java/com/yyxnb/yyxarch/livedata/ObservableLceeLiveData.kt
package com.yyxnb.yyxarch.livedata

import android.arch.lifecycle.LiveData
import android.support.annotation.NonNull
import com.yyxnb.yyxarch.bean.Lcee
import com.yyxnb.yyxarch.http.RetrofitManager
import com.yyxnb.yyxarch.http.exception.ApiException
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import java.lang.ref.WeakReference

internal class ObservableLceeLiveData<T>(private val mObservable: Observable<T>) : LiveData<Lcee<T>>() {
    private var mDisposableRef: WeakReference<Disposable>? = null
    private val mLock = Any()

    override fun onActive() {
        super.onActive()

        mObservable.subscribe(object : Observer<T> {
            override fun onSubscribe(@NonNull d: Disposable) {
                // Don't worry about backpressure. If the stream is too noisy then
                // backpressure can be handled upstream.
                synchronized(mLock) {
                    mDisposableRef = WeakReference(d)
                    RetrofitManager.addDisposable(mDisposableRef!!.get())
                }
                postValue(Lcee.loading())
            }

            override fun onNext(@NonNull t: T) {
                if (null == t) {
                    postValue(Lcee.empty())
                } else {
                    postValue(Lcee.content(t))
                }
            }

            override fun onError(@NonNull t: Throwable) {
                synchronized(mLock) {
                    RetrofitManager.cancelSingleRequest(mDisposableRef!!.get())
                    mDisposableRef = null
                }
                // Errors should be handled upstream, so propagate as a crash.
                postValue(Lcee.error(ApiException.handleException(t).message!!))
            }

            override fun onComplete() {
                synchronized(mLock) {
                    RetrofitManager.cancelSingleRequest(mDisposableRef!!.get())
                    mDisposableRef = null
                }
            }
        })

    }

    override fun onInactive() {
        super.onInactive()
        synchronized(mLock) {
            val subscriptionRef = mDisposableRef
            if (subscriptionRef != null) {
                val subscription = subscriptionRef.get()
                subscription?.dispose()
                RetrofitManager.cancelSingleRequest(mDisposableRef!!.get())
                mDisposableRef = null
            }
        }
    }
}
